### PR TITLE
feat(ci)!: add a reusable workflow for deployable integration tests

### LINKED ISSUE #3881: Old PRs can add a new "latest" cached state w/ the old state version
## Motivation

We don't check the state version when updating the cached state, so old PRs can add a new "latest" cached state with the old state version. This won't break CI, but it adds 2.5 hours to the mergify queue.

We can check cached state version vs the version in the current branch being tested. Doing this requires making the cached state version easier to find in CI. One idea is adding the state version as disk image metadata or in the name of the disk, and then doing some Rust test / grep check to compare the latest cached state disk version with the in-branch version, and only using the cached state if they are the same version. 

If the cached state version is higher than the branch version, we can not run the test, failing the test until the branch is rebased onto #main.

If the cached state version is lower than the branch version, we can run the test from an empty disk? Requires a full sync to the stop point.

Either way, the goal is that 'old' state versions (older than the version of state on #main) should not be created and pushed to gcloud, where they can be used by other tests.

### LINKED ISSUE #4239: ci: refactor test workflows to use a DRY/matrix approach
## Motivation

We have a bunch of workflow code repeating to execute almost the same tasks, for example the following jobs have over a 100 line of codes, and both are almost a **replica**:
- `test-full-sync`
- `test-stateful-sync`

This also happens with:
- `test-all`
- `test-fake-activation-heights`
- `test-empty-sync`
- `test-lightwalletd-integration`
- `regenerate-stateful-disks`

This same behavior is repeating with:
https://github.com/ZcashFoundation/zebra/pull/4223

### Designs

- As most test doesn't share a standard cli execution, this makes pipeline standardization a bit harder
- Using `entrypoint.sh` can workaround most of this issues. And might be the ideal solution until a more standardize way of executing the test is created within zebra/rust/cargo.
- All tests can have an execution variable with an specific nomenclature

## Related Work

https://github.com/ZcashFoundation/zebra/pull/4237

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

We have a bunch of workflow code repeating to execute almost the same tasks, for example the following jobs have over a 100 line of codes, and both are almost a **replica**:
- `test-full-sync`
- `test-stateful-sync`

This also happens with:
- `test-all`
- `test-fake-activation-heights`
- `test-empty-sync`
- `test-lightwalletd-integration`
- `regenerate-stateful-disks`

This is making the pipeline very hard to review, and it's not helping the team understand what's going on, or even being able to add new tests without having a deep understanding of how the pipeline works.

Fixes: #4239
Fixes: #3881

## Solution

- Create a reusable workflow which can be used for all tests that need to be deployed to GCP, and consider the multiple scenarios available

## Review

This is a kind of complex PR to understand, the whole @ZcashFoundation/devops-reviewers should have a look at this one.

### Reviewer Checklist

  - [ ] Validate the variables (inputs) being sent
  - [ ] Confirm the test execution in `entrypoint.sh`
  - [ ] Tests for Errors

## Follow Up Work

Add a matrix for:
- `test-all`
- `test-fake-activation-heights`
- `test-empty-sync`
- `test-lightwalletd-integration`
- `regenerate-stateful-disks`

And include VARIABLES for all of those in `entrypoint.sh`

### CODE REVIEW

**Review by teor2345 — COMMENTED** [review]
It looks like this PR diff contains changes from other PRs, so I'll review it once they are merged.

File: .github/workflows/gcp-test-deploy.yml
Comment (teor2345) [inline .github/workflows/gcp-test-deploy.yml]: Bug: `reusable workflow cannot be nested`

File: .github/workflows/gcp-test-deploy.yml
Comment (teor2345) [inline .github/workflows/gcp-test-deploy.yml]: Bug: the syntax in this line fails validation, probably due to nested expressions: `${{ inputs.${{ inputs.test_id }}s`

File: .github/workflows/gcp-test-deploy.yml
Comment (teor2345) [inline .github/workflows/gcp-test-deploy.yml]: Bug: missing end quote

```suggestion
          ${{ env.GAR_BASE }}/${{ env.IMAGE_NAME }}:sha-${{ env.GITHUB_SHA_SHORT }}"
```

File: .github/workflows/gcp-test-deploy.yml
Comment (teor2345) [inline .github/workflows/gcp-test-deploy.yml]: Minor: `STATE_VERSION appears unused`

```suggestion
```

File: .github/workflows/gcp-test-deploy.yml
Comment (teor2345) [inline .github/workflows/gcp-test-deploy.yml]:
```suggestion
          ${{ env.GAR_BASE }}/${{ env.IMAGE_NAME }}:sha-${{ env.GITHUB_SHA_SHORT }}"
```

File: docker/entrypoint.sh
Comment (teor2345) [inline docker/entrypoint.sh]: Possible Bug: duplicate commas in feature list

```suggestion
            exec cargo "test" "--locked" "--release" "--features" "enable-sentry,test_sync_to_mandatory_checkpoint_${NETWORK }" "--manifest-path" "zebrad/Cargo.toml" "sync_to_mandatory_checkpoint_${NETWORK }"
        elif [[ "$TEST_CHECKPOINT_SYNC" -eq "1" ]]; then
            exec cargo "test" "--locked" "--release" "--features" "enable-sentry,test_sync_past_mandatory_checkpoint_${NETWORK }" "--manifest-path" "zebrad/Cargo.toml" "test_sync_past_mandatory_checkpoint_${NETWORK }"
```

Comment (teor2345) [inline docker/entrypoint.sh]: Note: caching will be more efficient when we change from Rust features to environmental variables for these tests.

Until then, we just re-compile the test and `zebrad` binaries, which isn't too bad.

**Review by teor2345 — COMMENTED** [review]
I'd like to review this again once it has been tested.

File: docker/entrypoint.sh
Comment (gustavovalverde) [reply docker/entrypoint.sh]: Fixed by: [fix(rw): minor fixes](https://github.com/ZcashFoundation/zebra/pull/4271/commits/044e944715ff59bccd0eb2f415f9a4737cfb336a)

File: .github/workflows/gcp-test-deploy.yml
Comment (gustavovalverde) [reply .github/workflows/gcp-test-deploy.yml]: Fixed by: [fix(rw): minor fixes](https://github.com/ZcashFoundation/zebra/pull/4271/commits/044e944715ff59bccd0eb2f415f9a4737cfb336a)

File: .github/workflows/gcp-test-deploy.yml
Comment (gustavovalverde) [reply .github/workflows/gcp-test-deploy.yml]: Fixed by: [fix(rw): minor fixes](https://github.com/ZcashFoundation/zebra/pull/4271/commits/044e944715ff59bccd0eb2f415f9a4737cfb336a)

File: .github/workflows/gcp-test-deploy.yml
Comment (gustavovalverde) [reply .github/workflows/gcp-test-deploy.yml]: Fixed by: [fix(rw): minor fixes](https://github.com/ZcashFoundation/zebra/pull/4271/commits/044e944715ff59bccd0eb2f415f9a4737cfb336a)

File: .github/workflows/gcp-test-deploy.yml
Comment (gustavovalverde) [reply .github/workflows/gcp-test-deploy.yml]: Fixed by: [fix(rw): minor fixes](https://github.com/ZcashFoundation/zebra/pull/4271/commits/044e944715ff59bccd0eb2f415f9a4737cfb336a)

File: .github/workflows/gcp-test-deploy.yml
Comment (gustavovalverde) [reply .github/workflows/gcp-test-deploy.yml]: Fixed by: [fix(rw): remove nested worflow](https://github.com/ZcashFoundation/zebra/pull/4271/commits/3b6c87cf0cbdd99e2fccf9d07721208471f7e790)

**Comment by gustavovalverde** [comment]
@teor2345 tests are running. Let's see how it goes.

**Comment by teor2345** [comment]
Sync past checkpoint failed with:
> invalid argument "=1" for "-e, --env" flag: invalid environment variable: =1

**Comment by teor2345** [comment]
The main branch is failing with:
> docker: Error response from daemon: invalid mount config for type "bind": bind source path does not exist: /mnt/disks/gce-containers-mounts/gce-persistent-disks/zebrad-cache-9b9578c-mainnet-checkpoint.

> "docker run" requires at least 1 argument.

So we can't merge this PR until they are fixed.

File: .github/workflows/gcp-test-deploy.yml
Comment (teor2345) [reply .github/workflows/gcp-test-deploy.yml]: This change is not in that commit, please re-apply.

**Comment by teor2345** [comment]
@mergifyio update

File: .github/workflows/gcp-test-deploy.yml
Comment (gustavovalverde) [reply .github/workflows/gcp-test-deploy.yml]: It's being used here https://github.com/ZcashFoundation/zebra/pull/4271/files#diff-b62c23498eced92972377fdfbaf08c0d1518cabd782645cca9bfc2bfa12711edR144

**Review by dconnolly — COMMENTED** [review]
The actionlinter is giving a lot of feedback on gcp-test-deploy.yml

File: .github/workflows/test.yml
Comment (dconnolly) [inline .github/workflows/test.yml]: I think skew between the test id which is the source of truth about what is being run and the test name could be confusing, maybe we skip including a test_name and instead of all human-readable display names we can have `Run ${{ inputs.test_id }} test`

**Review by dconnolly — DISMISSED** [review]

File: .github/workflows/test.yml
Comment (dconnolly) [reply .github/workflows/test.yml]: I'm less opinionated about the `test_description` field

File: .github/workflows/test.yml
Comment (dconnolly) [inline .github/workflows/test.yml]: This is a pretty great pattern 🎉

File: .github/workflows/test.yml
Comment (dconnolly) [inline .github/workflows/test.yml]: This seems common for all right now, is there a way to have a default value in the `workflow_call` inputs that will allow this to be `/zebrad-cache` by default but optionally override-able? I think you can do similar with `workflow_dispatch`

Comment (dconnolly) [reply .github/workflows/test.yml]: Similar for `disk_prefix`

Comment (dconnolly) [reply .github/workflows/test.yml]: Similar for `app_name`

Comment (gustavovalverde) [reply .github/workflows/test.yml]: Great idea! Yes, this can be done.

https://docs.github.com/en/enterprise-cloud@latest/actions/using-workflows/workflow-syntax-for-github-actions#onworkflow_callinputs

File: .github/workflows/gcp-test-deploy.yml
Comment (dconnolly) [inline .github/workflows/gcp-test-deploy.yml]: This might need to be changed as this may not run a `sync` depending on the test it actually executes

File: .github/workflows/gcp-test-deploy.yml
Comment (dconnolly) [inline .github/workflows/gcp-test-deploy.yml]: Maybe `save_disk` ?

File: .github/workflows/gcp-test-deploy.yml
Comment (dconnolly) [inline .github/workflows/gcp-test-deploy.yml]:
```suggestion
  test-with-cached-state:
    name: Test using cached state
```

**Review by dconnolly — COMMENTED** [review]
This is looking really good!

File: .github/workflows/test.yml
Comment (gustavovalverde) [reply .github/workflows/test.yml]: Applied

File: .github/workflows/test.yml
Comment (gustavovalverde) [reply .github/workflows/test.yml]: Applied

File: .github/workflows/gcp-test-deploy.yml
Comment (gustavovalverde) [reply .github/workflows/gcp-test-deploy.yml]: Applied

File: .github/workflows/gcp-test-deploy.yml
Comment (gustavovalverde) [reply .github/workflows/gcp-test-deploy.yml]: Applied

File: .github/workflows/test.yml
Comment (teor2345) [inline .github/workflows/test.yml]: I wouldn't want to rely on the exact format here:
```suggestion
      height_grep_text: 'finished initial sync to chain tip, using gossiped blocks sync_percent=100.* current_height=Height'
```

File: .github/workflows/gcp-test-deploy.yml
Comment (teor2345) [reply .github/workflows/gcp-test-deploy.yml]: It's a bit confusing, but:
- `LOCAL_STATE_VERSION` is used in this step
- `STATE_VERSION` is created in `$GITHUB_ENV` for the next step, but not actually used in this step

You could rename `LOCAL_STATE_VERSION` to `STATE_VERSION`, that would fix this issue.

**Review by teor2345 — COMMENTED** [review]
Looks good, just a few minor issues left.

**Comment by gustavovalverde** [comment]
All issues identified while running the 3 tests have been fixed.

I'll assume this PR will get merged, so I'll inherit it to continue with `lightwalletd` inclusion into CI, as it's hard to modify all the dispersed code from the previous implementation, and considering most of `lightwalletd` integration tests require this same approach.

cc: @dconnolly @teor2345

File: .github/workflows/gcp-test-deploy.yml
Comment (teor2345) [reply .github/workflows/gcp-test-deploy.yml]: I'm just going to apply this fix, because it seems to have been missed twice.

**Comment by dconnolly** [comment]
> All issues identified while running the 3 tests have been fixed.
> 
> I'll assume this PR will get merged, so I'll inherit it to continue with `lightwalletd` inclusion into CI, as it's hard to modify all the dispersed code from the previous implementation, and considering most of `lightwalletd` integration tests require this same approach.
> 
> cc: @dconnolly @teor2345

@gustavovalverde  Can we fix the actionlint errors on here before we merge?

**Comment by gustavovalverde** [comment]
@dconnolly this are the errors we're getting:

`"description" is missing at "test_description" input of workflow_call event [syntax-check]` -> this one should be suggestion, as the `workflow_call` does not require a description for each output, that's our call if we want to add a description to each one.

`shellcheck reported issue in this script: SC2086:info:2:37: Double quote to prevent globbing and word splitting [shellcheck]` -> I don't think this is fixable

`shellcheck reported issue in this script: SC2140:warning:4:71: Word is of the form "A"B"C" (B indicated). Did you mean "ABC" or "A\"B\"C"? [shellcheck]` -> I don't understand this one

**Comment by teor2345** [comment]
I don't think we need to fix any of the remaining shellcheck warnings.

> `shellcheck reported issue in this script: SC2086:info:2:37: Double quote to prevent globbing and word splitting [shellcheck]` -> I don't think this is fixable

We can fix this one by quoting all the environmental variables we use, or disabling the check when we really do want to glob.

But it is advisory for us, because our environmental variables don't contain spaces or wildcards.

> `shellcheck reported issue in this script: SC2140:warning:4:71: Word is of the form "A"B"C" (B indicated). Did you mean "ABC" or "A\"B\"C"? [shellcheck]` -> I don't understand this one

We have arguments like: `--arg=config="$VAR",config2="$VAR2"`

But shells allow: `"--arg=config=$VAR,config2=$VAR2`, which is the canonical form, and sometimes easier to read.

But it can also accidentally change the variable names, so I don't think we should fix it.

**Comment by dconnolly** [comment]
> I don't think we need to fix any of the remaining shellcheck warnings.
> 
> > `shellcheck reported issue in this script: SC2086:info:2:37: Double quote to prevent globbing and word splitting [shellcheck]` -> I don't think this is fixable
> 
> We can fix this one by quoting all the environmental variables we use, or disabling the check when we really do want to glob.
> 
> But it is advisory for us, because our environmental variables don't contain spaces or wildcards.
> 
> > `shellcheck reported issue in this script: SC2140:warning:4:71: Word is of the form "A"B"C" (B indicated). Did you mean "ABC" or "A\"B\"C"? [shellcheck]` -> I don't understand this one
> 
> We have arguments like: `--arg=config="$VAR",config2="$VAR2"`
> 
> But shells allow: `"--arg=config=$VAR,config2=$VAR2`, which is the canonical form, and sometimes easier to read.
> 
> But it can also accidentally change the variable names, so I don't think we should fix it.

Can we configure actionlint to make these warnings and not hard errors, or something?

**Comment by teor2345** [comment]
> > I don't think we need to fix any of the remaining shellcheck warnings.
> > > `shellcheck reported issue in this script: SC2086:info:2:37: Double quote to prevent globbing and word splitting [shellcheck]` -> I don't think this is fixable
> > 
> > 
> > We can fix this one by quoting all the environmental variables we use, or disabling the check when we really do want to glob.
> > But it is advisory for us, because our environmental variables don't contain spaces or wildcards.
> > > `shellcheck reported issue in this script: SC2140:warning:4:71: Word is of the form "A"B"C" (B indicated). Did you mean "ABC" or "A\"B\"C"? [shellcheck]` -> I don't understand this one
> > 
> > 
> > We have arguments like: `--arg=config="$VAR",config2="$VAR2"`
> > But shells allow: `"--arg=config=$VAR,config2=$VAR2`, which is the canonical form, and sometimes easier to read.
> > But it can also accidentally change the variable names, so I don't think we should fix it.
> 
> Can we configure actionlint to make these warnings and not hard errors, or something?

Yes, but improving lints is not on the critical path, so we can do it later.

**Review by teor2345 — DISMISSED** [review]
Alright, let's see how this goes...

**Comment by gustavovalverde** [comment]
Last fixes applied @teor2345 

If we find something else, we'll do it in another PR.

**Review by teor2345 — APPROVED** [review]

**Pull request merged**
